Complete bounty pallet. Stages finished - "work submission", "judgment", "withdrawal".

shamil-gadelshin opened the pull request:
This PR completes the bounty pallet creation work.

### Changes
- limits for minimal values of cherry and funding.
- removed initial creator funding
- allow the council to fund the bounty
- new extrinsics:
  - announce_work_entry
  - withdraw_work_entry
  - submit_work
  - submit_oracle_judgment
  - withdraw_work_entrant_funds
- closed assurance contract checks
- major rework of the bounty stage calculation
- rewards for winners
- unstaking for winners and "legitimate participants"
- fine-grained "invalid stage" errors
- changed crate structure - new modules
    - `actors.rs` for bounty actor management 
    - `stages.rs` for bounty stage calculation

Supersedes this [PR](https://github.com/Joystream/joystream/pull/2218)

###  Open questions
_1. if we just store winners and to_be_slashed, and we do no iteration even on
these sets, but instead require active action from everyone, then we may perhaps not even need to
limit the set size? We could store thousands of ids in either of those values without impacting
read io speed, as we have seen, so could be fine._

_2. Should we have exactly 100% reward distribution for the winners?
 We could allow less than 100% but what to do with the rest of the bounty pool?
 It could be a problem because the remaining sum could be very low to be an incentive to withdraw for funders.
 Should we introduce a remaining sum threshold that affects the "burning decision"?_

[Main issue](https://github.com/Joystream/joystream/issues/1998)

shamil-gadelshin committed:
runtime: bounty: Add announce_work_entry extrinsic.

shamil-gadelshin committed:
runtime: bounty: Add announce_work_entry() tests.

shamil-gadelshin committed:
runtime: bounty: Add benchmarking for announce_work_entry() extrinsic.

shamil-gadelshin committed:
runtime: bounty: Add withdraw_work_entry extrinsic.

shamil-gadelshin committed:
runtime: bounty: Add MaxWorkEntryLimit.

shamil-gadelshin committed:
runtime: bounty: Add stake slashing on the work entry withdrawal .

shamil-gadelshin committed:
runtime: bounty: Introduce a min cherry value limit.

shamil-gadelshin committed:
runtime: bounty: Add minimum funding limit.

shamil-gadelshin committed:
runtime: bounty: Fix benchmarks.

shamil-gadelshin committed:
runtime: bounty: Rename BountyCreator to BountyActor

shamil-gadelshin committed:
runtime: bounty: Fix tests.

shamil-gadelshin committed:
runtime: bounty: Add more benchmarks.

Add benchmarks for coucil funding and funds withdrawal.

shamil-gadelshin committed:
runtime: bounty: Add more tests.

shamil-gadelshin committed:
runtime: bounty: Remove creator funding.

shamil-gadelshin committed:
runtime: bounty: Add benchmark for withdraw_work_entry.

Modify weight functions for fund_bounty, withdraw_funding, withdraw_work_entry.

shamil-gadelshin committed:
runtime: bounty: Add submit_work() extrinsic.

shamil-gadelshin committed:
runtime: bounty: Minor refactoring.

- rename error message
- remove obsolete TODOs

shamil-gadelshin committed:
Merge branch 'bounty' into bounty_work_submission

# Conflicts:
#	runtime-modules/bounty/src/benchmarking.rs

shamil-gadelshin committed:
runtime: bounty: Change a comment.

shamil-gadelshin committed:
runtime: bounty: Add closed contract check for work entries.

shamil-gadelshin committed:
runtime: bounty: Add closed contract checks on bounty creation.

shamil-gadelshin committed:
runtime: bounty: Add submit_oracle_judgement() extrinsic.

shamil-gadelshin committed:
runtime: bounty: Refactor bounty state and stages.

shamil-gadelshin committed:
runtime: bounty: Refactor bounty stage.

shamil-gadelshin committed:
runtime: bounty: Simplify bounty stage calculations.

shamil-gadelshin committed:
runtime: bounty: Improve tests.

shamil-gadelshin committed:
runtime: bounty: Add tests for submit_ judgement().

shamil-gadelshin committed:
runtime: bounty: Add more tests for submit_judgement()

shamil-gadelshin committed:
runtime: bounty: Fix judgment naming.

shamil-gadelshin committed:
runtime: bounty: Add benchmarks for submit_oracle_judgment()

shamil-gadelshin committed:
runtime: bounty: Refactor benchmarking.

shamil-gadelshin committed:
runtime: bounty: Fix PR comments.

shamil-gadelshin committed:
runtime: bounty: Add fine-grained stage errors.

shamil-gadelshin committed:
runtime: bounty: Introduce withdraw_work_entrant_funds() extrinsic.

shamil-gadelshin committed:
runtime: bounty: Refactor crate structure.

- move BountyActorManager to the separate actors.rs file.
- move BountyStageCalculator to the separate stage.rs file.

shamil-gadelshin committed:
runtime: bounty: Introduce work entry rewards.

shamil-gadelshin committed:
runtime: bounty: Check that a staking account belongs to a member.

shamil-gadelshin committed:
runtime: bounty: Update pallet comments.

shamil-gadelshin committed:
runtime: bounty: Add judgment winner reward checks.

shamil-gadelshin committed:
runtime: bounty: Add some tests for withdraw_work_entrant_funds()

shamil-gadelshin committed:
runtime: bounty: Update bounty removing.

shamil-gadelshin committed:
runtime: bounty: Add more tests for withdraw_work_entrant_funds().

shamil-gadelshin committed:
runtime: bounty: Add benchmark for withdraw_work_entrant_funds()

shamil-gadelshin commented:
#### Answering "Open questions"
1. We need to iterate through the "winners" to calculate and check the total winners' reward. We also have a limited "closed contract members" number.  So, "judgment" collection size should be limited as well.

2. We have several options:
   - demand the 100% funds distribution on the judgment,
   - burn undistributed funds (**current implementation**),
   - allow non-100% funds distribution on the judgment and allow funders to withdraw the rest. This option assumes that we don't worry about the state bloat by non-removed bounties left by funders (because of possible too low remaining funds).

shamil-gadelshin committed:
runtime: bounty: Clean TODO comments.

shamil-gadelshin committed:
runtime: bounty: Fix benchmarking PR comments.

conectado reviewed (approved):
Benchmarks LGTM

bedeho reviewed (changes requested):
Excellent work wrapping this, up, I think I significantly underestimated the complexity of this scope, but you have managed it well here.

Leaving a list of points here, most of which we have discussed offline.

# Conceptual

1. Make staking > 0 for all entries, and thus providing a staking account is not optional.

2. Judgements only cover entries that are to be paid (winners) or to be slashed (malicious), and there is a hardcap on the total number. Use map, not set! Judgement does cleanup for these and creator in successful bounty, so no action is required by these users. Everyone else, i.e. honest losers, are not referenced directly by judgements.

3. Remove generalized cherry reclaiming by creator, replace it by automated cleanup in 2 + a new _stage_ called `FundingExpired` which is when non-perpetual funding period expired AND no funder contributed anything. Expand `cancel_bounty` to also work in this stage, so that creator can reclaim cherry. Remove explicit `withdraw_creator_cherry`.

4. Split up generalized withdrawl stage into withdrawl stages for failed or successful bounty.

5. Do not allow over funding, if funding exceeds max limit, then only use required amount and leave funder the difference.

For 2,3,4, the following state chart summarizes functional organization of stages

![bounty_module](https://user-images.githubusercontent.com/437292/110990258-60983e80-8373-11eb-9c90-3c96f3a4df2c.png)


# Practical

6. Flesh out event payloads.

7.  Reject min, max = 0 funding limits, and enf

8.  flatten WorkEntries storage map to just be single key map.

9. drop `WorkEntryRecord::last_submitted_work`.

10. drop `WorkEntryRecord::reward_account_id` and just use controller account of member

11. rename `WorkEntry` to just `Entry`.

12. there are some mutables floating around, are they truly worth it? revist.

13. Vague comment: The `BountyActorManager` type appears to have vague meaning and role, although its clearly useful in some way to do some repetitive tasks. Is there some way to sharpen this? Even the idea of a bounty actor concept is a broad, becuase it does not cover the worker role.

shamil-gadelshin committed:
runtime: bounty: Make work entrant stake mandatory.

- Refactoring: rename “WorkEntry” type to “Entry”.

shamil-gadelshin committed:
runtime: bounty: Change KeepAlive to AllowDeath on transfers.

shamil-gadelshin committed:
runtime: bounty: Remove separate reward account.

Replace reward account with member controller account.

shamil-gadelshin committed:
runtime: bounty: Introduce FundingType

- FundingType defines funding conditions for perpetual and limited funding.

shamil-gadelshin committed:
runtime: bounty: Add FundingExpired bounty stage.

- allow canceling on FundingExpired

shamil-gadelshin committed:
runtime: Refactor common pallet.

- move types to a separate membership and council modules
- remove the origin pallet

shamil-gadelshin committed:
runtime: bounty: Automate cherry withdrawal on judgment.

bedeho commented:
# Updated State Chart

![bounty_module(1)](https://user-images.githubusercontent.com/437292/111768094-c753b580-88a7-11eb-986b-5f7b104dc36f.png)

shamil-gadelshin committed:
runtime: bounty: Remove withdraw_creator_cherry() extrinsic

- modify Withdrawal bounty stage (remove cherry_needs_withdrawal field)
- remove CreatorCherryWithdrawn bounty milestone

shamil-gadelshin committed:
runtime: bounty: Improve BountyCreated event.

shamil-gadelshin committed:
runtime: bounty: Add work entrant stake minimum limit.

shamil-gadelshin committed:
runtime: bounty: Demand 100% bounty distribution for winners.

shamil-gadelshin committed:
runtime: bounty: Fix benchmarks.

shamil-gadelshin committed:
runtime: bounty: Demand exact maximum funding amount.

shamil-gadelshin committed:
runtime: bounty: Change veto_ and cancel_bounty() extrinsics.

- add auto-returning of the creator cherry
- add bounty removal

shamil-gadelshin committed:
runtime: bounty: Add zero work entries check for judgement.

shamil-gadelshin committed:
runtime: bounty: Add work submission check for winners.

- Replace `last_submitted_work` data hash with `work_submitted` flag.

shamil-gadelshin committed:
runtime: bounty: Remove max work entry limit.

shamil-gadelshin committed:
runtime: bounty: Fix benchmarks.

shamil-gadelshin committed:
runtime: bounty: Change Entries to simple storage map.

- Change Entries storage from a double map to a simple map.

shamil-gadelshin committed:
runtime: bounty: Split Withdrawal stage into two separate stages.

Split Withdrawal stage into SuccessfulBountyWithdrawal and FailedBountyWithdrawal.

shamil-gadelshin commented:
### Review comments fixes summary

#### Major changes
- make a stake for work entries mandatory
- remove `withdraw_creator_cherry` extrinsic
- auto-transfer cherry on successful bounty judgment, or bounty canceling (veto)
- remove a bounty on cancellation or veto
- split `Withdrawal` stage into separate `SuccessfulBountyWithdrawal` and `FailedBountyWithdrawal`
- introduce exact funding for a maximum amount of limited funding and the target value for a perpetual one
- change "burn the rest" algorithm to the "demand 100% distribution" on bounty judgment
- remove work entries number limit
#### Minor changes
- introduced a limit for a minimum stake
- improve the event for `create_bounty` extrinsic
- introduce a special `FundingType` with separate parameters for limited and perpetual fundings.
- introduce a minimum funding limit.
- change `Entries` storage from a double map to a simple map
- change `EntryRecord::last_submitted_work` to work_submitted flag
- remove `EntryRecord::reward_account_id` 
- rename `WorkEntry` to `Entry`
- add "non-zero work entries number" check for the `submit_oracle_judgment` extrinsic
- add "entry has work submission" check for the `submit_oracle_judgment` extrinsic
- changes for benchmarks and tests.

shamil-gadelshin committed:
runtime: bounty: Renew weights for create_bounty() extrinsic.

bedeho reviewed (changes requested):
Phenomenal work here, just oustanding, only minor fixes

1. ` Amount of stake required, possibly 0, to enter bounty as entrant.`
2. Drop `OracleWorkEntryJudgment::Legit´, and instead make `EntryRecord::oracle_judgment_result` into an `Option`, this avoids judgements specifying legit outcomes, as well as having to set oracle judgement results that are not true.

shamil-gadelshin committed:
runtime: bounty: Remove “Legit” type for oracle judgment.

- make oracle judgment type optional for the work entry
- update comments for bounty creation parameters (cherry, entrant_stake)

shamil-gadelshin commented:
#### Next changeset
- removed `Legit` oracle judgment type for a work entry
- changed obsolete comment for `BountyCreationParameters::entrant_stake`
- changed obsolete comment for `BountyCreationParameters::cherry`

bedeho reviewed (approved):
Great!

I will leave benchmarking review to @conectado.

conectado commented on runtime-modules/bounty/src/benchmarking.rs:
```diff
@@ -80,51 +93,150 @@ fn member_funded_account<T: Trait + membership::Trait>(
 
         T::CouncilBudgetManager::set_budget(cherry);
 
         let params = BountyCreationParameters::<T>{
```
I think this should also have a `Closed` contract type since it's the worst case scenario for `ensure_create_bounty_parameters_valid`.

shamil-gadelshin replied:
OK.

conectado replied:
I think it'd be a good idea to have a `let j in 1 .. T::ClosedContractSizeLimit::get()` for the number of members, we can use that to calculate the weight in the extrinsic.

conectado commented on runtime-modules/bounty/src/benchmarking.rs:
```diff
@@ -134,32 +246,37 @@ benchmarks! {
+            creator: BountyActor::Member(member_id),
+            funding_type: FundingType::Perpetual{ target: max_amount },
             ..Default::default()
         };
```
I think this should also have a `Closed` contract type since it's the worst case scenario for `ensure_create_bounty_parameters_valid`.

shamil-gadelshin replied:
OK.

conectado reviewed (changes requested):
Benchmarks look mostly good, just 2 very small comments.

shamil-gadelshin committed:
runtime: bounty: Change create_bounty benchmark.

- Change create_bounty benchmark to the worst case scenario of the closed assurance contract.

conectado reviewed (commented):
Just 1 small thing

shamil-gadelshin committed:
runtime: bounty: Change create_bounty benchmark.
